addons/google_account/models/google_service.py

```diff
@@ -134,32 +135,46 @@ def _do_request(self, uri, params=None, headers=None, method='POST', preuri="htt
             headers = {}
 
         _logger.debug("Uri: %s - Type : %s - Headers: %s - Params : %s !", uri, method, headers, params)
-
-        ask_time = fields.Datetime.now()
-        try:
-            if method.upper() in ('GET', 'DELETE'):
-                res = requests.request(method.lower(), preuri + uri, params=params, timeout=timeout)
-            elif method.upper() in ('POST', 'PATCH', 'PUT'):
-                res = requests.request(method.lower(), preuri + uri, data=params, headers=headers, timeout=timeout)
-            else:
-                raise Exception(_('Method not supported [%s] not in [GET, POST, PUT, PATCH or DELETE]!') % (method))
-            res.raise_for_status()
-            status = res.status_code
-
-            if int(status) in (204, 404):  # Page not found, no response
-                response = False
-            else:
-                response = res.json()
-
+        # Implement exponential backoff as suggested by google documentation
+        # see https://developers.google.com/calendar/v3/errors#403_rate_limit_exceeded
+        retry = 1
+        while True:
+            ask_time = fields.Datetime.now()
             try:
-                ask_time = datetime.strptime(res.headers.get('date'), "%a, %d %b %Y %H:%M:%S %Z")
-            except:
-                pass
-        except requests.HTTPError as error:
-            if error.response.status_code in (204, 404):
-                status = error.response.status_code
-                response = ""
-            else:
-                _logger.exception("Bad google request : %s !", error.response.content)
-                raise error
+                if method.upper() in ('GET', 'DELETE'):
+                    res = requests.request(method.lower(), preuri + uri, params=params, timeout=timeout)
+                elif method.upper() in ('POST', 'PATCH', 'PUT'):
+                    res = requests.request(method.lower(), preuri + uri, data=params, headers=headers, timeout=timeout)
+                else:
+                    raise Exception(_('Method not supported [%s] not in [GET, POST, PUT, PATCH or DELETE]!') % (method))
+                res.raise_for_status()
+                status = res.status_code
+
+                if int(status) in (204, 404):  # Page not found, no response
```

Comment:
404 is impossible here, because of the `raise_for_status()` 2 lines above. (not your code, I know, just mentioning in passing)